File: Samples/Win7Samples/sysmgmt/windowspowershell/csharp/getprocesssample05/GetProcessSample05.cs
//
// Copyright (c) 2006 Microsoft Corporation. All rights reserved.
// 
// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF 
// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A 
// PARTICULAR PURPOSE.
//
using System;
using System.Collections.Generic;
using Win32Exception = System.ComponentModel.Win32Exception;
using System.Diagnostics;
using System.Security.Permissions;
using System.Management.Automation;    // Windows PowerShell namespace


// This sample is a complete implementation of the get-proc cmdlet.
//
// To test this cmdlet, create a module folder that has the same name 
// as this assembly, and then run the following command:
// import-module getprocesssample05

namespace Microsoft.Samples.PowerShell.Commands
{
    #region GetProcCommand

    /// <summary>
   /// This class implements the get-proc cmdlet.
   /// </summary>
   [Cmdlet(VerbsCommon.Get, "Proc", 
      DefaultParameterSetName="ProcessName")]
   public class GetProcCommand: PSCmdlet
   {
      #region Parameters

      /// <summary>
      /// This parameter provides the list of process names on 
      /// which the Get-Proc cmdlet will work.
      /// </summary>
      [Parameter(
         Position = 0,
         ParameterSetName = "ProcessName",
         ValueFromPipeline = true,
         ValueFromPipelineByPropertyName = true)]
      [ValidateNotNullOrEmpty]
      public string[] Name
      {
         get { return processNames; }
         set { processNames = value; }
      }

      /// <summary>
      /// This parameter provides the list of process identifiers on 
      /// which the Get-Proc cmdlet will work.
      /// </summary>
      [Parameter(
         ParameterSetName = "Id",
         Mandatory = true,
         ValueFromPipeline = true,
         ValueFromPipelineByPropertyName = true,
         HelpMessage = "The unique id of the process to get.")]
      public int[] Id
      {
         get { return processIds; }
         set { processIds = value; }
      }

      /// <summary>
      /// This parameter takes Process objects directly. If the input is a 
      /// stream of [collection of] Process objects, the cmdlet bypasses the 
      /// ProcessName and Id parameters and reads the Process objects 
      /// directly.  This allows the cmdlet to deal with processes that have 
      /// wildcard characters in their name.
      /// <value>Process objects</value>
      /// </summary>
      [Parameter(
         ParameterSetName = "InputObject",
         Mandatory = true,
         ValueFromPipeline = true)]
      public Process[] Input
      {
         get { return inputObjects; }
         set { inputObjects = value; }
      }

      #endregion Parameters

      #region Cmdlet Overrides

      /// <summary>
      /// The ProcessRecord method calls the Process.GetProcesses 
      /// method to retrieve the processes. Then, the WriteObject 
      /// method writes the associated processes to the pipeline.
      /// </summary>
      protected override void ProcessRecord()
      {
         List<Process> matchingProcesses;
         
         WriteDebug("Obtaining the list of matching process objects.");

         switch (ParameterSetName)
         {
            case "Id":
               matchingProcesses = GetMatchingProcessesById();
               break;
            case "ProcessName":
               matchingProcesses = GetMatchingProcessesByName();
               break;
            case "InputObject":
               matchingProcesses = GetProcessesByInput();
               break;
            default:
               ThrowTerminatingError(
                   new ErrorRecord(
                       new ArgumentException("Bad ParameterSetName"),
                       "UnableToAccessProcessList",
                       ErrorCategory.InvalidOperation,
                       null));
               return;
         } // switch (ParameterSetName)

         WriteDebug("Outputting the matching process objects.");

         matchingProcesses.Sort(ProcessComparison);

         foreach (Process process in matchingProcesses)
         {
            WriteObject(process);
         }
      } // ProcessRecord

      #endregion Overrides

      #region protected Methods and Data

      /// <summary>
      /// Retrieves the list of all processes matching the ProcessName
      /// parameter and generates a nonterminating error for each 
      /// specified process name which is not found even though the name
      /// contains no wildcards.
      /// </summary>
      /// <returns></returns>
      [EnvironmentPermissionAttribute(
         SecurityAction.LinkDemand, 
         Unrestricted = true)]
      private List<Process> GetMatchingProcessesByName()
      {
         new EnvironmentPermission(
            PermissionState.Unrestricted).Assert();
         
         List<Process> allProcesses = 
            new List<Process>(Process.GetProcesses());

         // The keys dictionary is used for rapid lookup of 
         // processes that are already in the matchingProcesses list.
         Dictionary<int, byte> keys = new Dictionary<int, byte>();

         List<Process> matchingProcesses = new List<Process>();

         if (null == processNames)
            matchingProcesses.AddRange(allProcesses);
         else
         {
            foreach (string pattern in processNames)
            {
               WriteVerbose("Finding matches for process name \""
                  + pattern + "\".");

               // WildCard serach on the available processes
               WildcardPattern wildcard =
                  new WildcardPattern(pattern, 
                     WildcardOptions.IgnoreCase);

               bool found = false;

               foreach (Process process in allProcesses)
               {
                  if (!keys.ContainsKey(process.Id))
                  {
                     string processName = SafeGetProcessName(process);

                     // Remove the process from the allProcesses list 
                     // so that it is not tested again.
                     if(processName.Length == 0)
                        allProcesses.Remove(process);

                     // Perform a wildcard search on this particular 
                     // process name and check whether it matches the 
                     // pattern specified.
                     if (!wildcard.IsMatch(processName))
                        continue;

                     WriteDebug("Found matching process id " 
                        + process.Id + ".");

                     // A match is found.
                     found = true;
                     // Store the process identifier so that the same process
                     // is not added twice.
                     keys.Add(process.Id, 0);
                     // Add the process to the processes list.
                     matchingProcesses.Add(process);
                  }
               } // foreach (Process...

               if (!found &&
                 !WildcardPattern.ContainsWildcardCharacters(pattern))
               {
                  WriteError(new ErrorRecord(
                     new ArgumentException("Cannot find process name "
                        + "\"" + pattern + "\"."),
                     "ProcessNameNotFound",
                     ErrorCategory.ObjectNotFound,
                     pattern));
               }
            } // foreach (string...
         } // if (null...
         
         return matchingProcesses;
      } // GetMatchingProcessesByName

      /// <summary>
      /// Returns the name of a process.  If an error occurs, a blank
      /// string is returned.
      /// </summary>
      /// <param name="process">The process whose name is 
      /// returned.</param>
      /// <returns>The name of the process.</returns>
      [EnvironmentPermissionAttribute(
         SecurityAction.LinkDemand, Unrestricted = true)]
      protected static string SafeGetProcessName(Process process)
      {
         new EnvironmentPermission(PermissionState.Unrestricted).Assert();
         string name = "";

         if (process != null)
         {
            try 
            { 
                return process.ProcessName; 
            }
            catch (Win32Exception) { }
            catch (InvalidOperationException) { }
         }

         return name;
     } // SafeGetProcessName

      #endregion Cmdlet Overrides

      #region Private Methods

      /// <summary>
      /// Function to sort by process name first, and then by 
      /// the process identifier.
      /// </summary>
      /// <param name="x">First process object.</param>
      /// <param name="y">Second process object.</param>
      /// <returns>
      /// Returns less than zero if x is less than y,
      /// greater than 0 if x is greater than y, and 0 if x == y.
      /// </returns>
      private static int ProcessComparison(Process x, Process y)
      {
         int diff = String.Compare(
            SafeGetProcessName(x),
            SafeGetProcessName(y),
            StringComparison.CurrentCultureIgnoreCase);

         if (0 != diff)
            return diff;

         return x.Id - y.Id;
      }

      /// <summary>
      /// Retrieves the list of all processes matching the Id
      /// parameter and generates a nonterminating error for 
      /// each specified process identofier which is not found.
      /// </summary>
      /// <returns>An array of processes that match the given identifier.
      /// </returns>
      [EnvironmentPermissionAttribute(
         SecurityAction.LinkDemand,
         Unrestricted = true)]
      private List<Process> GetMatchingProcessesById()
      {
         new EnvironmentPermission(
            PermissionState.Unrestricted).Assert();

         List<Process> matchingProcesses = new List<Process>();

         if (null != processIds)
         {
            // The keys dictionary is used for rapid lookup of the 
            // processes already in the matchingProcesses list.
            Dictionary<int, byte> keys = new Dictionary<int, byte>();

            foreach (int processId in processIds)
            {
               WriteVerbose("Finding match for process id "
                  + processId + ".");

               if (!keys.ContainsKey(processId))
               {
                  Process process;
                  try { process = Process.GetProcessById(processId); }
                  catch (ArgumentException ex)
                  {
                     WriteError(new ErrorRecord(
                        ex,
                        "ProcessIdNotFound",
                        ErrorCategory.ObjectNotFound,
                        processId));
                     continue;
                  }
                  
                  WriteDebug("Found matching process.");
                  
                  matchingProcesses.Add(process);
                  keys.Add(processId, 0);
               }
            }
         }
         
         return matchingProcesses;
      } // GetMatchingProcessesById

      /// <summary>
      /// Retrieves the list of all processes matching the InputObject
      /// parameter.
      /// </summary>
      [EnvironmentPermissionAttribute(
         SecurityAction.LinkDemand,
         Unrestricted = true)]
      private List<Process> GetProcessesByInput()
      {
         new EnvironmentPermission(
            PermissionState.Unrestricted).Assert();

         List<Process> matchingProcesses = new List<Process>();

         if (null != Input)
         {
            // The keys dictionary is used for rapid lookup of the
            // processes already in the matchingProcesses list.
            Dictionary<int, byte> keys = new Dictionary<int, byte>();

            foreach (Process process in Input)
            {
               WriteVerbose("Refreshing process object.");

               if (!keys.ContainsKey(process.Id))
               {
                  try { process.Refresh(); }
                  catch (Win32Exception) { }
                  catch (InvalidOperationException) { }

                  matchingProcesses.Add(process);
               }
            }
         }
         return matchingProcesses;
      } // GetProcessesByInput

      #endregion Private Methods

      #region Private Data

      private string[] processNames;
      private int[] processIds;
      private Process[] inputObjects;

      #endregion Private Data

    } //GetProcCommand

    #endregion GetProcCommand
}

File: Samples/Win7Samples/sysmgmt/windowspowershell/vb/getprocesssample05/GetProcessSample05.vb
'
' Copyright (c) 2006 Microsoft Corporation. All rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF 
' ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
' THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A 
' PARTICULAR PURPOSE.
'

Imports System
Imports System.Collections.Generic
Imports Win32Exception = System.ComponentModel.Win32Exception
Imports System.Diagnostics
Imports System.Security.Permissions

'Windows PowerShell namespace
Imports System.Management.Automation
Imports System.ComponentModel

Namespace Microsoft.Samples.PowerShell.Commands

    ' This sample is a complete implementation of the get-proc Cmdlet.
#Region "GetProcCommand"

    ''' <summary>
    ''' This class implements the get-proc cmdlet
    ''' </summary>
    <Cmdlet(VerbsCommon.Get, "Proc", _
    DefaultParameterSetName:="ProcessName")> _
    Public Class GetProcCommand
        Inherits PSCmdlet
#Region "Parameters"

        ''' <summary>
        ''' The list of process names on which this cmdlet will work
        ''' </summary>
        <Parameter(Position:=0, ParameterSetName:="ProcessName", _
        ValueFromPipeline:=True, _
        ValueFromPipelineByPropertyName:=True), _
        ValidateNotNullOrEmpty()> _
        Public Property Name() As String()
            Get
                Return processNames
            End Get

            Set(ByVal value As String())
                processNames = value
            End Set

        End Property

        ''' <summary>
        ''' gets/sets an array of process IDs
        ''' </summary>
        <Parameter(ParameterSetName:="Id", _
        Mandatory:=True, ValueFromPipeline:=True, _
        ValueFromPipelineByPropertyName:=True, _
        HelpMessage:="The unique id of the process to get.")> _
        Public Property Id() As Integer()
            Get
                Return processIds
            End Get
            Set(ByVal value As Integer())
                processIds = value
            End Set
        End Property

        ''' <summary>
        ''' If the input is a stream of [collections of] Process 
        ''' objects, we bypass the ProcessName and Id parameters and 
        ''' read the Process objects directly.  This allows us to deal 
        ''' with processes which have wildcard characters in their name.
        ''' <value>Process objects</value>
        ''' </summary>
        <Parameter(ParameterSetName:="InputObject", _
        Mandatory:=True, ValueFromPipeline:=True)> _
        Public Property Input() As Process()
            Get
                Return inputObjects
            End Get
            Set(ByVal value As Process())
                inputObjects = value
            End Set
        End Property

#End Region

#Region "Cmdlet Overrides"

        ''' <summary>
        ''' For each of the requested processnames, retrieve and write
        ''' the associated processes.
        ''' </summary>
        Protected Overrides Sub ProcessRecord()
            Dim matchingProcesses As List(Of Process)

            WriteDebug("Obtaining list of matching process objects.")

            Select Case ParameterSetName
                Case "Id"
                    matchingProcesses = GetMatchingProcessesById()
                Case "ProcessName"
                    matchingProcesses = GetMatchingProcessesByName()
                Case "InputObject"
                    matchingProcesses = GetProcessesByInput()
                Case Else
                    ThrowTerminatingError(New ErrorRecord( _
                        New ArgumentException("Bad ParameterSetName"), _
                            "UnableToAccessProcessList", _
                            ErrorCategory.InvalidOperation, Nothing))
                    Return
            End Select

            WriteDebug("Outputting matching process objects.")

            matchingProcesses.Sort(AddressOf ProcessComparison)

            Dim process As Process
            For Each process In matchingProcesses
                WriteObject(process)
            Next process

        End Sub 'ProcessRecord

#End Region

#Region "protected Methods and Data"
        ''' <summary>
        ''' Retrieves the list of all processes matching the ProcessName
        ''' parameter.
        ''' Generates a non-terminating error for each specified
        ''' process name which is not found even though it contains
        ''' no wildcards.
        ''' </summary>
        ''' <returns></returns>

        Private Function GetMatchingProcessesByName() As List(Of Process)

            Dim allProcesses As List(Of Process) = _
                New List(Of Process)(Process.GetProcesses())

            ' The keys dictionary will be used for rapid lookup of 
            ' processes already in the matchingProcesses list.
            Dim keys As Dictionary(Of Integer, Byte) = _
                New Dictionary(Of Integer, Byte)()

            Dim matchingProcesses As List(Of Process) = New List(Of Process)()

            If Nothing Is processNames Then
                matchingProcesses.AddRange(allProcesses)
            Else
                Dim pattern As String
                For Each pattern In processNames
                    WriteVerbose(("Finding matches for process name """ & _
                        pattern & """."))

                    ' WildCard serach on the available processes
                    Dim wildcard As New WildcardPattern(pattern, _
                        WildcardOptions.IgnoreCase)

                    Dim found As Boolean = False

                    Dim process As Process
                    For Each process In allProcesses
                        If Not keys.ContainsKey(process.Id) Then
                            Dim processName As String = _
                                SafeGetProcessName(process)

                            ' Remove the process from the allProcesses list 
                            ' so that it's not tested again.
                            If processName.Length = 0 Then
                                allProcesses.Remove(process)
                            End If

                            ' Perform a wildcard search on this particular 
                            ' process and check whehter this matches the 
                            ' pattern specified.
                            If Not wildcard.IsMatch(processName) Then
                                GoTo ContinueForEach2
                            End If

                            WriteDebug(String.Format( _
                                "Found matching process id ""{0}"".", process.Id))

                            ' We have found a match.
                            found = True

                            ' Store the process ID so that we don't add the 
                            ' same one twice.
                            keys.Add(process.Id, 0)

                            ' Add the process to the processes list.
                            matchingProcesses.Add(process)
                        End If
ContinueForEach2:
                    Next process ' foreach (Process...
                    If Not found AndAlso Not _
                       WildcardPattern.ContainsWildcardCharacters(pattern) _
                    Then
                        WriteError(New ErrorRecord( _
                            New ArgumentException("Cannot find process name " & _
                            " " & pattern & "."), "ProcessNameNotFound", _
                            ErrorCategory.ObjectNotFound, pattern))
                    End If
                Next pattern
            End If
            Return matchingProcesses

        End Function 'GetMatchingProcessesByName

        ''' <summary>
        ''' Returns the name of a process.  If an error occurs, a blank
        ''' string will be returned.
        ''' </summary>
        ''' <param name="process">The process whose name will be 
        ''' returned.</param>
        ''' <returns>The name of the process.</returns>
        Protected Shared Function SafeGetProcessName(ByVal process As Process) _
            As String

            Dim name As String = ""

            If Not (process Is Nothing) Then
                Try
                    Return process.ProcessName
                Catch e1 As Win32Exception
                Catch e2 As InvalidOperationException
                End Try
            End If
            Return name


        End Function 'SafeGetProcessName

#End Region


#Region "Private Methods"

        ''' <summary>
        ''' Function to sort by ProcessName first, then by Id
        ''' </summary>
        ''' <param name="x">first Process object</param>
        ''' <param name="y">second Process object</param>
        ''' <returns>
        ''' returns less than zero if x less than y,
        ''' greater than 0 if x greater than y, 0 if x == y
        ''' </returns>
        Private Shared Function ProcessComparison(ByVal x As Process, _
                ByVal y As Process) As Integer
            Dim diff As Integer = String.Compare(SafeGetProcessName(x), _
            SafeGetProcessName(y), StringComparison.CurrentCultureIgnoreCase)

            If 0 <> diff Then
                Return diff
            End If
            Return x.Id - y.Id

        End Function 'ProcessComparison

        ''' <summary>
        ''' Retrieves the list of all processes matching the Id
        ''' parameter.
        ''' Generates a non-terminating error for each specified
        ''' process ID which is not found.
        ''' </summary>
        ''' <returns>An array of processes that match the given id.
        ''' </returns>
        Protected Function GetMatchingProcessesById() As List(Of Process)

            Dim matchingProcesses As List(Of Process) = New List(Of Process)

            If Not (processIds Is Nothing) Then

                ' The keys dictionary will be used for rapid lookup of 
                ' processes already in the matchingProcesses list.
                Dim keys As Dictionary(Of Integer, Byte) = _
                    New Dictionary(Of Integer, Byte)()

                Dim processId As Integer
                For Each processId In processIds
                    WriteVerbose("Finding match for process id " & _
                        processId & ".")

                    If Not keys.ContainsKey(processId) Then
                        Dim process As Process
                        Try
                            process = _
                               System.Diagnostics.Process.GetProcessById( _
                                   processId)
                        Catch ex As ArgumentException
                            WriteError(New ErrorRecord(ex, _
                                "ProcessIdNotFound", _
                                ErrorCategory.ObjectNotFound, processId))
                            GoTo ContinueForEach1
                        End Try

                        WriteDebug("Found matching process.")

                        matchingProcesses.Add(process)
                        keys.Add(processId, 0)
                    End If
ContinueForEach1:
                Next processId
            End If

            Return matchingProcesses

        End Function 'GetMatchingProcessesById

        ''' <summary>
        ''' Retrieves the list of all processes matching the Input
        ''' parameter.
        ''' </summary>
        Private Function GetProcessesByInput() As List(Of Process)

            Dim matchingProcesses As List(Of Process) = New List(Of Process)()

            If Not (Nothing Is Input) Then
                ' The keys dictionary will be used for rapid lookup of 
                ' processes already in the matchingProcesses list.
                Dim keys As Dictionary(Of Integer, Byte) = _
                    New Dictionary(Of Integer, Byte)()

                Dim process As Process
                For Each process In Input
                    WriteVerbose("Refreshing process object.")

                    If Not keys.ContainsKey(process.Id) Then
                        Try
                            process.Refresh()
                        Catch e1 As Win32Exception
                        Catch e2 As InvalidOperationException
                        End Try
                        matchingProcesses.Add(process)
                    End If
                Next process
            End If
            Return matchingProcesses

        End Function 'GetProcessesByInput

#End Region

#Region "Private Data"

        Private processNames() As String
        Private processIds() As Integer
        Private inputObjects() As Process

#End Region

    End Class 'GetProcCommand 

#End Region

#Region "PowerShell snap-in" '
    ''' <summary>
    ''' Create this sample as an PowerShell snap-in
    ''' </summary>
    <RunInstaller(True)> _
    Public Class GetProcPSSnapIn05
        Inherits PSSnapIn

        ''' <summary>
        ''' Create an instance of the GetProcPSSnapIn05
        ''' </summary>
        Public Sub New()

        End Sub 'New

        ''' <summary>
        ''' Get a name for this PowerShell snap-in. This name will
        ''' be used in registering
        ''' this PowerShell snap-in.
        ''' </summary>

        Public Overrides ReadOnly Property Name() As String
            Get
                Return "GetProcPSSnapIn05"
            End Get
        End Property

        ''' <summary>
        ''' Vendor information for this PowerShell snap-in.
        ''' </summary>

        Public Overrides ReadOnly Property Vendor() As String
            Get
                Return "Microsoft"
            End Get
        End Property

        ''' <summary>
        ''' Gets resource information for vendor. This is a string of format: 
        ''' resourceBaseName,resourceName. 
        ''' </summary>

        Public Overrides ReadOnly Property VendorResource() As String
            Get
                Return "GetProcPSSnapIn05,Microsoft"
            End Get
        End Property

        ''' <summary>
        ''' Description of this PowerShell snap-in.
        ''' </summary>

        Public Overrides ReadOnly Property Description() As String
            Get
                Return "This is a PowerShell snap-in that includes " & _
                    "the get-proc sample."
            End Get
        End Property
    End Class 'GetProcPSSnapIn05

#End Region

End Namespace
